### PR TITLE
feat(storage): inject failure in object_store

### PR DESCRIPTION
**Pull request by xxhZs**
## What's changed and what's your intention?

We inject failure in object_store(read,upload,delete), and write some ut.

1. Add some ut about mem_store failure(read,upload,delete);
2. Add regression tests for https://github.com/singularity-data/risingwave/issues/763 https://github.com/singularity-data/risingwave/pull/680
3. we add test_runner(thread_local) like tikv in stroage to avoid Fail_Point_Err in multi_threaded.
In stroage,  if the `ut.name` is contains `“test_failpoint”`,It will put FailScenario in thread-local.
## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests

### CODE REVIEW

File: src/storage/src/hummock/mod.rs [hunk 1]
```diff
@@ -581,7 +585,7 @@ impl StateStore for HummockStorage {
                         key_range.start_bound().map(|b| b.as_ref().to_owned()),
                     ),
                     epoch,
-                    Some(version),
+                    None,
```
Comment (hzxa21) [inline src/storage/src/hummock/mod.rs:588]: We need to pass the version to user iterator to ensure that the version is pinned when the iterator is valid. Why do we change it here?

Comment (xxhZs) [reply src/storage/src/hummock/mod.rs:588]: sorry , i change it for test in some commit , And `git merge` causes this errors . I will solve this problem. I am  So sorry

File: src/storage/src/hummock/iterator/test_utils.rs [hunk 2]
```diff
@@ -95,6 +95,28 @@ pub async fn gen_iterator_test_sstable_base(
     .await
 }
 
+pub async fn gen_iterator_test_sstable_base_without_buff(
```
Comment (MrCroxx) [inline src/storage/src/hummock/iterator/test_utils.rs:98]: `buff` -> `buf`? 🤣

File: src/storage/src/hummock/test_utils.rs [hunk 3]
```diff
@@ -92,21 +94,29 @@ pub fn gen_test_sstable_data(
...
+/// Generate a test table from the given `kv_iter` and put the kv value to `sstable_store`
+pub async fn gen_test_sstable(
+    opts: SSTableBuilderOptions,
+    sst_id: u64,
+    kv_iter: impl Iterator<Item = (Vec<u8>, HummockValue<Vec<u8>>)>,
+    sstable_store: SstableStoreRef,
+) -> Sstable {
+    gen_test_sstable_inner(opts, sst_id, kv_iter, sstable_store, CachePolicy::Fill).await
+}
+
```
Comment (MrCroxx) [inline src/storage/src/hummock/test_utils.rs:119]: Do we really need the wrapper? 🤔

File: src/storage/src/hummock/iterator/test_utils.rs (hunk 2 above)
Comment (xxhZs) [reply src/storage/src/hummock/iterator/test_utils.rs:98]: ok

File: src/storage/src/hummock/test_utils.rs (hunk 3 above)
Comment (xxhZs) [reply src/storage/src/hummock/test_utils.rs:119]: in this test, we need `CachePolicy::NotFill`,  So i use the wrapper (Most of the calling methods use the `CachePolicy::Fill`)

**Comment by MrCroxx** [comment]
**PLEASE NOTE THAT**, failpoints are not thread-safe, and failpoints will not be auto-disabled across tests.

By default, different test functions run concurrently, and failpoints will affect each other.

So, you need to set the parallelism of the tests with failpoints to 1. And that's the reason why TiKV separates the tests with failpoints into another directory.

@xxhZs @hzxa21

**Comment by MrCroxx** [comment]
... And TiKV uses its own test runner, so it doesn't config the parallelism in `Cargo.toml`. That's why you cannot find the parallelism config in the `Cargo.toml` of TiKV.

**Comment by xxhZs** [comment]
> **PLEASE NOTE THAT**, failpoints are not thread-safe, and failpoints will not be auto-disabled across tests.
> 
> By default, different test functions run concurrently, and failpoints will affect each other.
> 
> So, you need to set the parallelism of the tests with failpoints to 1. And that's the reason why TiKV separates the tests with failpoints into another directory.
> 
> @xxhZs @hzxa21

yes , we add test_runner(thread_local) like tikv in stroage to avoid the Testing conflicts(Normal test and failpoints test)

**Pull request merged**